Modules/mathmodule.c

```diff
@@ -2217,41 +2217,63 @@ math_modf_impl(PyObject *module, double x)
    in that int is larger than PY_SSIZE_T_MAX. */
 
 static PyObject*
-loghelper(PyObject* arg, double (*func)(double))
+loghelper_int(PyObject* arg, double (*func)(double))
 {
     /* If it is int, do it ourselves. */
-    if (PyLong_Check(arg)) {
-        double x, result;
-        Py_ssize_t e;
+    double x, result;
+    Py_ssize_t e;
 
-        /* Negative or zero inputs give a ValueError. */
-        if (!_PyLong_IsPositive((PyLongObject *)arg)) {
-            PyErr_SetString(PyExc_ValueError,
-                            "math domain error");
-            return NULL;
-        }
+    /* Negative or zero inputs give a ValueError. */
+    if (!_PyLong_IsPositive((PyLongObject *)arg)) {
+        PyErr_SetString(PyExc_ValueError,
+                        "math domain error");
+        return NULL;
+    }
 
-        x = PyLong_AsDouble(arg);
-        if (x == -1.0 && PyErr_Occurred()) {
-            if (!PyErr_ExceptionMatches(PyExc_OverflowError))
-                return NULL;
-            /* Here the conversion to double overflowed, but it's possible
-               to compute the log anyway.  Clear the exception and continue. */
-            PyErr_Clear();
-            x = _PyLong_Frexp((PyLongObject *)arg, &e);
-            if (x == -1.0 && PyErr_Occurred())
-                return NULL;
-            /* Value is ~= x * 2**e, so the log ~= log(x) + log(2) * e. */
-            result = func(x) + func(2.0) * e;
-        }
-        else
-            /* Successfully converted x to a double. */
-            result = func(x);
-        return PyFloat_FromDouble(result);
+    x = PyLong_AsDouble(arg);
+    if (x == -1.0 && PyErr_Occurred()) {
+        if (!PyErr_ExceptionMatches(PyExc_OverflowError))
+            return NULL;
+        /* Here the conversion to double overflowed, but it's possible
+           to compute the log anyway.  Clear the exception and continue. */
+        PyErr_Clear();
+        x = _PyLong_Frexp((PyLongObject *)arg, &e);
+        if (x == -1.0 && PyErr_Occurred())
+            return NULL;
+        /* Value is ~= x * 2**e, so the log ~= log(x) + log(2) * e. */
+        result = func(x) + func(2.0) * e;
     }
+    else
+        /* Successfully converted x to a double. */
+        result = func(x);
+    return PyFloat_FromDouble(result);
+}
 
+static PyObject*
+loghelper(PyObject* arg, double (*func)(double))
+{
+    /* If it is int, do it ourselves. */
+    if (PyLong_Check(arg)) {
+        return loghelper_int(arg, func);
+    }
```

Comment:
```suggestion
    /* If it is int or index-like, do it ourselves. */
    if (PyIndex_Check(arg)) {
        arg = _PyNumber_Index(arg);
        if (arg == NULL) {
            return NULL;
        }   
        PyObject *res = loghelper_int(arg, func);
        Py_DECREF(arg);
    }
```
How about this?  Tests pass with the rest reverted to ``return math_1(arg, func, 0)``.

Reply:
It is intentional. `__float__` has priority over `__index__` in `PyFloat_AsDouble()`. I do not want to change this here.

I did not bother to write a test for such weird case of `__float__` and `__index__` both defined and returning non-equal values, but if you insist, I can add it.

Reply:
> It is intentional. __float__ has priority over __index__ in PyFloat_AsDouble(). I do not want to change this here.

Perhaps, we could workaround this by checking also nb_float slot?  If it's not defined **and** PyIndex_Check is successful - take PyLong branch.  Else - fallback to math_1 and not try to use ``__index__()`` at all.

<details>

```suggestion
    PyNumberMethods *nb = Py_TYPE(arg)->tp_as_number;
    /* If it is int or index-like, do it ourselves. */
    if (nb != NULL && nb->nb_float == NULL && PyIndex_Check(arg)) {
        arg = _PyNumber_Index(arg);
        if (arg == NULL) {
            return NULL;
        }   
        PyObject *res = loghelper_int(arg, func);
        Py_DECREF(arg);
    }
```

</details>

> I did not bother to write a test for such weird case of __float__ and __index__ both defined and returning non-equal values

In fact, ``PyFloat_AsDouble()`` will not try ``__index__()`` dunder in this case.

> but if you insist, I can add it.

I'm +0 on this, but maybe it's wise to test, that ``__index__()`` dunder will be ignored in this case.  People sometimes do weird definitions for magic methods.  E.g. recently I've found that mpmath.mpf() type had ``__complex__()`` method.

Reply:
We still need to handle case when the type has both `__float__` and `__index__`, and `__float__` overflows.

So we can add the first part of your code, but keep the rest as is. But it will add a tiny overhead for every normal use case, so I am not sure that it is worth to do.

Reply:
> We still need to handle case when the type has both __float__ and __index__, and __float__ overflows.

Shouldn't we ignore ``__index__`` in this case as PyNumber_Float() does?

Reply:
I do not think so. `PyNumber_Float()` does not use `__index__` if `__float__` fails with overflow error, because the integer value cannot be represented as a floating point value. But its logarithm can, so it makes sense to use `__index__`.

I wonder if it makes sense to use `__int__`. The logarithm of the truncated values should be almost indistinguishable from the logarithm of the exact values. But that "almost" may play role for high-precision (more than 1024 bits of mantissa) numbers. The correction code may be too complex for this issue.

Reply:
> But its logarithm can, so it makes sense to use __index__.

But if there is ``__index__`` slot - it will be handled above in [my suggestion](https://github.com/python/cpython/pull/121011/files#r1659493860):)